src/goto-programs/validate_goto_modelt.cpp

```diff
@@ -0,0 +1,286 @@
+  result = check_returns_removed() && result;
+  result = check_called_functions() && result;
+  result = check_last_instruction() && result;
+
+  // Am aware of the true/false rtn debate. Currently use true for success/pass.
+  // I would like to prompt some feedback on method names above in case
+  // of returning false - e.g.
+  // a. preserve names but return opposite of implied
+  // b. or remove 'state' from name e.g. just use 'check_entry_point'
+  // c. or 'double negative' style e.g. 'entry_point_not_exists' - and return
+  // false when it does.
+  return result;
+}
+
+bool validate_goto_modelt::entry_point_exists()
+{
+  for(const auto &fun : function_map)
+  {
+    if(goto_functions.entry_point() == fun.first)
+      return true;
+  }
+  return false;
+}
+
+bool validate_goto_modelt::instruction_function_id_isvalid()
+{
+  for(const auto &fun : function_map)
+  {
+    for(auto &instr : fun.second.body.instructions)
+    {
+      if(instr.function != fun.first)
+        return false;
+    }
+  }
+  return true;
+}
+
+bool validate_goto_modelt::function_pointer_calls_removed()
+{
+  for(const auto &fun : function_map)
+  {
+    for(auto &instr : fun.second.body.instructions)
+    {
+      if(instr.is_function_call())
+      {
+        const code_function_callt &function_call =
+          to_code_function_call(instr.code);
+
+        if(function_call.function().id() == ID_dereference)
+          return false;
+      }
+    }
+  }
+  return true;
+}
+
+bool validate_goto_modelt::check_returns_removed()
+{
+  for(const auto &fun : function_map)
+  {
+    const goto_functiont &goto_function = fun.second;
+    for(const auto &instr : goto_function.body.instructions)
+    {
+      if(goto_function.type.return_type().id() != ID_empty)
+        return false;
+
+      if(instr.is_return())
+        return false;
+
+      if(instr.is_function_call())
+      {
+        const auto &function_call = to_code_function_call(instr.code);
+        if(!function_call.lhs().is_nil())
+          return false;
+
+        const auto &callee = to_code_type(function_call.function().type());
+        if(callee.return_type().id() != ID_empty)
+          return false;
+      }
+    }
+  }
+  return true;
+}
+
+bool validate_goto_modelt::check_called_functions()
+{
+  class test_for_function_addresst : public const_expr_visitort
+  {
+  private:
+    const exprt expr_address_of;
+
+  public:
+    bool found{false};
+    std::queue<irep_idt> identifiers; //preserve order
+
+    explicit test_for_function_addresst(exprt expr_address_of)
+      : expr_address_of{expr_address_of}
+    {
+    }
+
+    void operator()(const exprt &expr) override
+    {
+      if(expr.id() == ID_address_of)
+      {
+        auto &pointee = to_address_of_expr(expr).object();
+        if(pointee.type().id() == ID_code) // a function
+        {
+          found = true;
+          irep_idt identifier = to_symbol_expr(pointee).get_identifier();
```

Comment:
I think we also need to check here that the expression of which the address if taken is indeed a `symbol_exprt`. Technically in the code there could be constructs like `&(*fp)` where a function pointer is dereferenced and then the address is taken. We thus need something like:

```
if(pointee.id() == ID_symbol)
{
  identifiers.push(to_symbol_expr(pointee).get_identifier());
}
```